src/score.rs

```diff
@@ -1,34 +1,84 @@
 use std::cmp::min;
+use std::ascii::AsciiExt;
 
-pub fn score(choice: &String, query: &String) -> f32 {
+#[derive(Clone, Debug,PartialEq)]
+pub struct Quality(pub f32);
+
+impl Quality {
+    pub fn to_f32(&self) -> f32 {
+        let Quality(q) = *self;
+        q
+    }
+}
+
+#[derive(Clone, Debug,PartialEq)]
+pub struct Substring(pub usize,pub usize);
+
+impl Substring {
+    pub fn slice(&self, input: &String) -> (String, String, String) {
+       let Substring(start, end) = *self;
+       (input[0..start].to_string(),
+        input[start..end].to_string(),
+        input[end..].to_string())
+    }
+}
+
+#[derive(Clone, Debug,PartialEq)]
+pub struct Match<'a> {
+    pub quality: Quality,
+    pub range: Substring,
+    pub original: &'a String,
+
+}
+
+impl <'a>Match<'a>{
+    pub fn new(quality: Quality, range: Substring, original: &'a String) -> Match<'a> {
+        Match { quality: quality, range: range, original: original }
+    }
+
+    pub fn origin_only(original: &'a String) -> Match<'a> {
+        Match::new(Quality(1.0), Substring(0,0), original)
+    }
+}
+
+pub fn score<'a>(choice: &'a String, query: &String) -> Option<Match<'a>> {
     let choice_length = choice.len() as f32;
     let query_length = query.len() as f32;
 
-    if query_length == 0.0 { return 1.0 }
+    if query_length == 0.0 { return Some(Match::origin_only(choice)) }
+    let lower_choice = choice.to_ascii_lowercase();
 
-    match compute_match_length(choice, query) {
-        Some(match_length) => (query_length / match_length as f32) / choice_length,
-        None => 0.0,
+    match compute_match_length(&lower_choice, query) {
+        Some((start, match_length)) => {
+            let quality = Quality( (query_length / match_length as f32) / choice_length);
+            let substring = Substring(start, start+match_length);
```

Comment:
What do you think about pushing the substring logic to the components that are concerned with formatting and let search just return a match range without constructing a `Substring`? We should be able to leverage the built-in `Range` struct for that.

Something like `Match::new(quality, choice, range)`